### PR TITLE
Fixing #408 and #413

### PR DESCRIPTION
**Pull request by denizyuret**

### CODE REVIEW

File: test/header.jl [hunk 1]
```diff
@@ -1,2 +1,3 @@
 using Test, Random, AutoGrad, Knet
-include(AutoGrad.dir("test","gradcheck.jl"))
```
Comment (ekinakyurek) [inline test/header.jl]: why this is not enough `using AutoGrad: gradcheck, gcheck`

File: src/rnn.jl [hunk 2]
```diff
@@ -505,9 +518,15 @@ function rnnforw(r::RNN, w::KnetArray{T}, x::KnetArray{T},
     hyout = hyDesc = cyout = cyDesc = C_NULL
     if hy || cy
         firstBatchSize = batchSizes==nothing ? size(x,2) : batchSizes[1]
-        hsize = Int[r.hiddenSize, firstBatchSize, r.numLayers * (r.direction == 1 ? 2 : 1)] # (H,B,L/2L)
-        if hy; hyout=similar(y,hsize...); hyDesc=TD3(hyout); end
-        if cy && r.mode==2; cyout=similar(y,hsize...); cyDesc=TD3(cyout); end
+        hsize = (Int(r.hiddenSize), Int(firstBatchSize), Int(r.numLayers * (r.direction == 1 ? 2 : 1))) # (H,B,L/2L)
+        if hy; hyout=similar(y,hsize); hyDesc=TD3(hyout); end
+        if cy && r.mode==2; cyout=similar(y,hsize); cyDesc=TD3(cyout); end
+        if hx != nothing && any(size(hx,i)!=hsize[i] for i=1:3) # compare one by one in case hx is 1-D or 2-D
```
[the remark this answers is not in the document]
Comment (ekinakyurek) [reply src/rnn.jl]: Could you share the thrown error when you call rnn forward?

[the remark this answers is not in the document]
Comment (ilkerkesen) [reply src/rnn.jl]: OK, let me revert what I've done so far to fix this issue, then paste the stack trace...

**Comment by denizyuret** [comment]
The charlm tutorial example should have a working example usage.
The hiddens should be passed in an Any array as you said.
The Any array can either be empty, or contain appropriately sized hidden
KnetArray(s) (2 if lstm, 1 otherwise)
When you are done with the iteration, you should clean up using hidden .=
value.(hidden)
The new documentation should have this info. Let me know if this does not
help.

**Comment by ilkerkesen** [comment]
[Here](https://gist.github.com/ilkerkesen/f0e45c8048de2eec26e0604eadcc9669) is the error trace. @denizyuret I've just built documentation on my local machine, but I am not able to see any RNN documentation. Could you give me the URL path?

**Comment by ekinakyurek** [comment]
I believe the bug is due to this [line](https://github.com/denizyuret/Knet.jl/blob/e972ed6ef03dd5dfd33179f63a92de212d8928fa/src/rnn.jl#L137) and the line after. Although, your hiddens are empty, these two line forces hy=cy=true. It causes the error later. I believe this bug can be solved easily @denizyuret.

However, what you want is that: collect hidden states without giving initial values of them. I am afraid that cudnn doesn't allow this. So, I mean you can't do this even in the old interface.

You can give zero arrays as initial hidden states. Then, your problem will be solved.

File: src/rnn.jl [hunk 3]
```diff
@@ -127,7 +136,8 @@ function (r::RNN)(x; batchSizes=nothing, hidden=nothing)
     tr = !isempty(AutoGrad._tapes)
     hy = (hidden != nothing)
```
Comment (ekinakyurek) [inline src/rnn.jl]: if hidden is empty, this causes bug later

**Comment by ekinakyurek** [comment]
We may also initialize hiddens inside of the RNN function for users who want to collect hidden states.

**Comment by ilkerkesen** [comment]
> However, what you want is that: collect hidden states without giving initial values of them. I am afraid that cudnn doesn't allow this. So, I mean you can't do this even in the old interface.

We are able to do this in the old interface.

> You can give zero arrays as initial hidden states. Then, your problem will be solved.

Construction zero arrays inside loss function does not hurt the tape, does it? I am not sure, so I've avoided it so far.

**Comment by ekinakyurek** [comment]
Yes, I check it out you could do it before. Sorry for the misinformation.

I believe, you can construct zero arrays both inside and outside of the loss without ruining your gradient calculation.

**Pull request merged**